pkg/containerruntimediscovery/dockerdiscoverer.go

```diff
@@ -0,0 +1,147 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containerruntimediscovery
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	dtypes "github.com/docker/docker/api/types"
+	dclient "github.com/docker/docker/client"
+	"github.com/sirupsen/logrus"
+
+	"github.com/openclarity/vmclarity/api/models"
+	"github.com/openclarity/vmclarity/pkg/shared/log"
+	"github.com/openclarity/vmclarity/pkg/shared/utils"
+)
+
+type DockerDiscoverer struct {
+	client *dclient.Client
+	logger *logrus.Entry
+}
+
+func NewDockerDiscoverer(ctx context.Context) (Discoverer, error) {
+	client, err := dclient.NewClientWithOpts(dclient.FromEnv, dclient.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create docker client: %w", err)
+	}
+
+	_, err = client.ImageList(ctx, dtypes.ImageListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list images: %w", err)
+	}
+
+	return &DockerDiscoverer{
+		client: client,
+		logger: log.GetLoggerFromContextOrDefault(ctx),
+	}, nil
+}
+
+func (dd *DockerDiscoverer) Images(ctx context.Context) ([]models.ContainerImageInfo, error) {
+	// List all docker images
+	images, err := dd.client.ImageList(ctx, dtypes.ImageListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list images: %w", err)
+	}
+
+	// Convert images to container image info
+	result := make([]models.ContainerImageInfo, len(images))
```

Comment:
Why do you not use the merge logic as you did in containerd discoverer?

Reply:
Docker natively groups images by ImageID, containerd doesn't and stores a separate entry per manifest ID so in ContainerD we have to merge the entries together by ImageID outselves.